transit.md

```diff
@@ -217,3 +272,108 @@ both sides have their own nonce.
+hint. The following section is closely tied to the operating system's sockets API, and the
+the verbs *bind*, *listen*, *accept* and *connect* will be used accordingly. Note that this
+is more a guide than a specification, since it handles implementation details that should
+all mostly be compatible with each other (and when not, failure is not critical).
+
+### The easy way
+
+Each side binds a new TCP socket, enumerates all local non-loopback IP addresses and exchanges
+port+IPs with the peer. Then, both sides listen on their socket while simultaneously trying to
+connect to the peer. As described above, once a handshake is successful, all other attempts are
+aborted and the resources closed.
+
+### Punching holes through firewalls
+
+To go through firewalls, a different approach must be used instead: first, both sides again need
+to bind a socket and exchange hints. This time however, the `SO_REUSEADDR` option must be set.
+Then, both sides try to connect to each of the peer's IPs. For every connection attempt, they bind
+a new socket on the same port again. Also you read that right, there are no `listen` or `accept`
+calls involved.
+
+This works only with firewalls that silently filter packets and which don't send back any RST packets.
+Unfortunately, this is the default behavior of the kernel when no firewall is active (or the firewall allowed)
+the packets. This means that the firewall hole punching method does *not* work when *no* firewall is
+active (Fun fact: if one side has a firewall and the other side doesn't then connection attempts will
+work or fail depending on which side sent their packets first—I'm really glad for you that you didn't
+have to debug this …). Thus, **this method can only be an addition to the "easy" way** specified above.
+Sadly, this means that the number of hints – and thus, attempted connections – is doubled. If you manage
+to listen and accept on the same port so that only one set of hints needs to be sent out, please contact
+`piegames` with your findings. Also read [this paper](https://bford.info/pub/net/p2pnat/) (Python code
+[here](https://github.com/dwoz/python-nat-hole-punching/)) for more details on the subject.
+
+### Traversing NAT
+
+To additionally traverse through a potential NAT, clients need to query their external IP address
+from a STUN server\*. Because of how NATs work, the socket must again be bound with `SO_REUSEADDR`
+and its port must be used for the hint. Clients may `shutdown` the connection to the STUN server, but
+the socket must be kept open. No enhanced NAT detection or other advanced STUN features need to be
+used: we only care about the external IP address and if it fails we can always simply fall back to
+the relay. Generally, a failure to do STUN should be silent and the query should time out rather
+quickly.
+
+Recall that NATs are a structure of the legacy and that IPv6 users usually have a globally routed
```

Comment:
"of the legacy ..." word missing here? (I guess you mean "Internet" maybe?). I'm not sure that dismissing "the IPv4 Internet" as "legacy" helps the cause here; it'll be around for some time by the looks of things.